### ISSUE TITLE
Improve msfvenom apk evasion

### ISSUE THREAD

**Issue by busterb** [rapid7/metasploit-framework#8034]
The apkwash script does some things to bypass fragile av signatures, could be integrated into apk generation directly:

https://github.com/jbreed/apkwash/blob/master/apkwash

https://www.youtube.com/watch?v=z3tPSBo9E_k

**Comment by timwr** [comment]
Great stuff @jbreed
We already do some of those things when we inject into an existing APK (with msfvenom -x), but that script looks like it applies it to the default payload.
It's going to be a little fiddly to support those techniques for the default payload APK. We will need aapt (which is bundled in apktool, I think) to parse AndroidManifest.xml and the resources, and baksmali/smali (or metasm?) to parse the smali in classes.dex file (also included in apktool). 
TL/DR: Maybe it's easier we keep the default payload static (and not add a dependency on apktool), and just add those techniques to -x?

**Comment by jbreed** [comment]
That would be awesome to have something integrated directly into the framework with the mindset to force AV companies to get ahead of the curve. It is sad when it took me a couple hours to manually find all of their signatures, then another day or so to write-up a script for it. 

For simply wanting a very small/portable payload like the default payload, this seemed to work best. I wish I would have known about the '-x' feature before making 'https://github.com/jbreed/apkinjector' as I'm sure the framework would make it a lot easier and work more often. I found apktool to fail pulling all the smali files on larger apps like Facebook, Starbucks, etc.

**Comment by timwr** [comment]
msfvenom -x for apks has moved on a lot recently. I've incorporated some more randomisation into #8038.
I'm happy to keep playing cat and mouse with AV vendors that rely on signatures, feel free to open up an issue here if you find one that can detect it.
@SkullTech I noticed you have an awesome article on injecting into APKs, however it might be worth updating it to say it's now supported with `msfvenom -x`

**Comment by dana-at-cp** [comment]
I am a heavy advocate for pushing mobile security vendors to do better. In my backdoor-apk project I use evasion techniques similar to those of apkwash, msfvenom -x etc. but I also use simple obfuscation of "const-string" values in payload related smali files.

https://github.com/dana-at-cp/backdoor-apk

I inject a custom Java class that is required to unobfuscate the "const-string" values and then automagically modify the smali files to make the static method calls where necessary. The custom Java class is included in a Proguard processing step so I think it's generally immune from being used in any meaningful mobile antivirus signatures itself. The same is true for the static method calls added in the smali files. I think any mobile antivirus signatures targeting those method calls would be prone to false positive.

With that said, I could explore porting over this functionality to 'msfvenom -x' or devising a different (perhaps better) scheme.

**Comment by timwr** [comment]
@dana-at-cp obfuscating the strings would be great. I've removed most of them but there are one or two left over. We can wait and see what is flagged next.
Obfuscating strings could also be useful because it allows us to use reflection to obfuscate the code further.
Can you provide the source code for StringObfuscator.class? So it replaces the const-string values with static method calls into that class, which decode and return the strings value?

**Comment by timwr** [comment]
I can't decide if there is a better fix for https://github.com/rapid7/metasploit-payloads/pull/175/files
I've moved it to be placed inside the config now (for stageless payloads only), but I wonder if I could somehow convince proguard to strip it out of staged payloads, but keep it in stageless.

**Comment by dana-at-cp** [comment]
@timwr You got it. I have some code in the shell script that rewrites const-string values in specific smali files. The StringObfuscator.class includes a static method that can undo the obfuscation. I should mention that the obfuscation mechanism is a simple caesar cipher so eventually it will need to be replaced with something more sophisticated. This was merely to prove that the concept was viable.

Here's a sample of the log output from the shell script:

```
StringObfuscator class: net/dial/cust1/e
File: /root/Code/github/backdoor-apk/backdoor-apk/original/smali/net/dial/cust1/AppBoot.smali
Target: "android.intent.action.BOOT_COMPLETED"
Reg: v0
Shift count: 18
Shift tr value: [I-ZA-Hi-za-h]
Replacement: "18ivlzwql.qvbmvb.ikbqwv.JWWB_KWUXTMBML"
File: /root/Code/github/backdoor-apk/backdoor-apk/original/smali/net/dial/cust1/a.smali
Target: "com.metasploit.meterpreter.AndroidMeterpreter"
Reg: v1
Shift count: 18
Shift tr value: [I-ZA-Hi-za-h]
Replacement: "18kwu.umbiaxtwqb.umbmzxzmbmz.IvlzwqlUmbmzxzmbmz"
File: /root/Code/github/backdoor-apk/backdoor-apk/original/smali/net/dial/cust1/a.smali
Target: "payload.jar"
Reg: v2
Shift count: 20
Shift tr value: [G-ZA-Fg-za-f]
Replacement: "20vgerugj.pgx"
File: /root/Code/github/backdoor-apk/backdoor-apk/original/smali/net/dial/cust1/a.smali
Target: "payload.dex"
Reg: v3
Shift count: 22
Shift tr value: [E-ZA-De-za-d]
Replacement: "22tecpseh.hib"
...
```

The shell script replaces a const-string declaration, such as:

```
     const-string v1, "com.metasploit.meterpreter.AndroidMeterpreter"
````

With three new lines that look like this:

```
    const-string v1, "18kwu.umbiaxtwqb.umbmzxzmbmz.IvlzwqlUmbmzxzmbmz"

    invoke-static {v1}, Lnet/dial/cust1/e;->a(Ljava/lang/String;)Ljava/lang/String;

    move-result-object v1
```

I'll attach the Java source file soon.
